[FE] 새로운 이슈 작성 페이지

youzysu opened the pull request:
## Issues
- #41 
- #51 
- #52 
- #73

## What is this PR? 👓
- 새로운 이슈 작성 페이지 UI 및 주요 기능
- TextArea, Sidebar, ProgressBar 컴포넌트 구현
- CircleCheckbox: input type checkbox으로 RadioInput과 UI는 동일

## Key changes 🔑
- Sidebar 컴포넌트 구현
- Sidebar 컴포넌트에서 DropdownIndicator 재사용을 위해 Dropdownitem 단일 선택/복수 선택 option에 대한 prop 추가
  - 단일 선택/복수 선택 & context 유무에 따라 맞는 Input Element 렌더링 (checkbox / radio type)
- Sidebar 부모 컴포넌트와 dropdownItem 선택 상태 공유를 위해 checkboxContext, radioContext 생성
  - 복수 선택 가능한 Assignee, Labels는 CheckboxGroup, 단일 선택인 Milestone은 RadioGroup 사용
- DropdownItem item props id 추가
- TextInput 컴포넌트 개선: focused 상태 삭제
- 기타 변수명 수정 (loginId -> username 등)

## To reviewers 👋
- Dropdown은 컴포넌트 크기나 재사용성 고려해서 이후에 개선이 필요해보여요. 지금처럼 props로 Dropdown의 세부 조건을 전달해주고, Dropdown 내부에서 모두 결정하기보다는 prop으로 variant만 받고 하위에서 variant에 따라 아예 다른 컴포넌트를 리턴하도록 하는 식으로 이후에 개선해보면 어떨까 싶네요! 제 생각에는 아예 드롭다운 종류를 여러개로 두는게 더 낫다고 생각이 들어요. 구체적인 개선 방향에 대해서는 4주차 때 더 고민해보면 좋을 듯 해요.
- Depth가 깊어서 Context를 사용하다보니 구조가 살짝 복잡해보이나 싶네요, 다른 해결 방법이나 의견 주세요! 구현하면서 제가 고민한 포인트는 이슈에 작성해뒀어요.
- ProgressBar는 이슈 상세 조회 페이지 UI는 아직 고려하지 않아서 개선해야 해요.
- 작업 단위가 전보다 좀 커서 놓친 부분이 있을 수도 있지만, 카카의 꼼꼼한 리뷰를 믿습니다 ㅎㅎ 🙏

youzysu committed:
style: Button Component XL 사이즈 추가

- 로그인/회원가입 버튼

youzysu committed:
refactor: Label, Avatar 컴포넌트 재사용을 위한 분리

youzysu committed:
#51 feat: TextArea Component UI

youzysu committed:
#52 feat: Sidebar Component 기본 UI, mock API & data 분리 및 getUser 추가

youzysu committed:
#41 feat: 새로운 이슈 작성 페이지 UI 구현

Kakamotobi commented on fe/src/components/common/Sidebar.tsx:
```diff
@@ -0,0 +1,131 @@
+}) {
+  const userList = useFetch<User[]>([], getUsers);
+  const allLabels = useFetch<Label[]>([], getLabels);
+  const milestonesList = useFetch<Milestone[]>([], getMilestones);
```
이런 경우에 "all"로 통일해도 좋을 것 같아요. `allUsers`, `allMilestones`.

Kakamotobi commented on fe/src/components/common/Sidebar.tsx:
```diff
@@ -0,0 +1,131 @@
+        </div>
+      );
+    });
+  };
```
`generateIssueAssigneeList`, `generateIssueLabelList`라고 하면 덜 헷갈릴것 같아요

youzysu replied:
지금은 각 컴포넌트로 분리된 상태입니다

Kakamotobi commented on fe/src/components/common/TextArea.tsx:
```diff
@@ -0,0 +1,109 @@
+  value,
+  ...props
+}: React.TextareaHTMLAttributes<HTMLTextAreaElement>) {
+  const [isFocused, setIsFocused] = useState(false);
```
`body` 계속 flash되는거 해결됐나요?

Kakamotobi commented on fe/src/components/common/TextArea.tsx:
```diff
@@ -0,0 +1,109 @@
+        {...props}
+      />
+      <FileUploadButton>
+        <label className="file-upload-label" htmlFor="file-upload-input">
```
만들어두신 `Label` 컴포넌트를 여기에도 사용하면 될 것 같습니다.

Kakamotobi commented on fe/src/pages/MainPage/NewIssuePage.tsx:
```diff
@@ -0,0 +1,134 @@
+    setContent(targetValue);
+  };
+
+  const moveToMainPage = () => navigate("/");
+
+  const onIssueSubmit = async () => {
```
순서 바꿔주시면 더 깔끔해보일 것 같습니다

Kakamotobi commented on fe/src/pages/MainPage/NewIssuePage.tsx:
```diff
@@ -0,0 +1,134 @@
+  }
+`;
+
+const Title = styled.span`
```
제목은 semantic하게 h 태그가 좋을 것 같아요

Kakamotobi commented on fe/src/pages/MainPage/NewIssuePage.tsx:
```diff
@@ -0,0 +1,134 @@
+  color: ${({ theme: { neutral } }) => neutral.text.strong};
+`;
+
+const Line = styled.div`
```
`Line` 컴포넌트 대신 그냥 `.wrapper`에 `border-top`, `border-bottom`을 입혀도 되지 않을까요?

Kakamotobi commented on fe/src/pages/MainPage/NewIssuePage.tsx:
```diff
@@ -0,0 +1,134 @@
+  flex-direction: column;
+  gap: 24px;
+
+  .line {
+    border-bottom: ${({ theme: { border, neutral } }) =>
+      `${border.default} ${neutral.border.default}`};
+  }
```
안쓰는 코드인것 같은데 맞나요?

youzysu committed:
#52 feat: Sidebar 컴포넌트 일부 구현

- AddAssignee, AddLabel 구현
- DropdownIndicator 컴포넌트 재사용을 위해 DropdownItem 컴포넌트 수정
- CircleCheckbox 컴포넌트 생성
- 선택된 담당자, 라벨 상태 관리를 위해 CheckGroup, CheckContext 생성
- Type 수정

youzysu committed:
#41 feat: 새로운 이슈 추가 기능

- 새로운 이슈 추가 페이지
- mock API & data 추가

youzysu committed:
rename: loginId에서 username으로 수정

youzysu committed:
#51 refactor: input focus 상태 제거 & css :focus로 스타일 구현

youzysu committed:
fix: DropdownList type error 해결

- 타입 명시

youzysu committed:
#41 feat: 이미지 업로드 후 링크 마크다운 형식으로 추가

- 선택한 이미지 업로드 요청
- 이미지 링크 Content에 추가

youzysu committed:
#73 design: ProgressBar 컴포넌트 UI

youzysu committed:
#52 feat, refactor: Sidebar 컴포넌트 구조 개선, RadioContext & RadioGroup 생성

- 마일스톤은 단일 선택으로 RadioContext & RadioGroup 필요
- Group 컴포넌트가 DropdownIndicator만 포함하도록 개선 & fieldset 태그로 수정
- Dropdown option props 추가: 다중선택 / 단일선택 옵션을 의미
- DropdownItem에서 option에 따라 다른 타입의 Input 렌더링 (경우의 수 4가지 가능)
- 기타 네이밍 수정

youzysu committed:
#41 feat: 새로운 이슈 마일스톤 설정 기능

youzysu committed:
refactor #41: 새로운 이슈 정보 상태 수정

- newIssueInfo 객체로 변경

youzysu committed:
style: TODO 주석 작성 & 불필요한 주석 삭제

- TODO: labels, milestone, assignees optional 처리 필요

youzysu committed:
#51 feat: TextArea 추가 기능

- 글자수 보여주는 기능
- TextArea 스크롤바 및 dashed 디자인 추가

Kakamotobi commented on fe/src/components/common/Sidebar/AddAssignee.tsx:
`AddAssignee`보다 `AssigneeField`같은 이름이 더 명확한 것 같습니다.

Kakamotobi commented on fe/src/components/common/Sidebar/AddAssignee.tsx:
```diff
@@ -0,0 +1,79 @@
+  assignees: number[];
+  onAssigneeChange: (assignees: number[]) => void;
+}) {
+  const userList = useFetch<User[]>([], getUsers);
+
+  const userMap: UserMap = userList.reduce((map: UserMap, user: User) => {
+    map[user.userAccountId] = user;
+    return map;
+  }, {});
+
+  const assigneeDropdownList: DropdownItemType[] = userList.map((user) => ({
+    id: user.userAccountId,
+    variant: "withImg",
+    name: "assignee",
+    content: user.username,
+    imgSrc: user.profileUrl,
+  }));
+
+  const generateAssignees = () => {
+    return assignees.map((assignee) => {
+      const currentAssignee = userMap[assignee];
+      return (
+        <Wrapper key={currentAssignee.userAccountId}>
+          <Avatar
+            src={currentAssignee.profileUrl}
+            alt={`${currentAssignee.username}`}
+            $size="S"
+          />
+          <Content>{currentAssignee.username}</Content>
+        </Wrapper>
+      );
+    });
+  };
```
`newIssueInfo.assignees`를 배열이 아닌 Set으로 두면 `Set.has()`으로 로직이 더 깔끔해질 것 같습니다. 
Ex: `userList.filter()` 해서 `assignees.has(user.userAccountId)`이 true면, 선택된 assignee 만들기.

Kakamotobi commented on fe/src/components/common/Sidebar/CheckboxGroup.tsx:
`CheckboxGroup`이랑 `RadioGroup`은 Sidebar 뿐만 아니라 다른데서도 사용 가능하기 때문에 `Sidebar` 폴더 밖으로 `common`에 두어도 될 것 같습니다

youzysu committed:
#51 refactor: TextArea 글자수 보여주기 위한 useEffect 로직 개선

- isCharCountShown 상태 변경 함수 외부로 분리

youzysu committed:
#41 #52 refactor: Sidebar 컴포넌트 개선 & 변경사항 새로운 이슈 페이지 반영

- 담당자, 레이블 상태 자료구조 Set으로 변경 및 로직 반영
- Group 컴포넌트 분리
- 컴포넌트 네이밍 수정: Field

youzysu committed:
#52 fix: MilestoneField 개선

- 불필요한 연산 제거
